File: mobile/lib/protocol/input_event.dart
import 'dart:convert';
import 'dart:typed_data';

import 'frame.dart';
import 'payload_codec.dart';

const int inputEventKindButton = 0x00;
const int inputEventKindAxis = 0x01;
const int inputEventKindStick = 0x02;
const int inputEventKindTrigger = 0x03;
const int inputEventKindHat = 0x04;

const int inputEventFlagStateChanged = 0x01;
const int inputEventFlagInitial = 0x02;

const int inputEventStateUp = 0x00;
const int inputEventStateDown = 0x01;

const int inputEventMinControlIdLength = 1;
const int inputEventMaxControlIdLength = 64;

class InputEvent {
  const InputEvent({
    required this.controlId,
    required this.kind,
    required this.flags,
    this.state,
    this.pressCount,
    this.value,
    this.x,
    this.y,
    this.hatValue,
  });

  final String controlId;
  final int kind;
  final int flags;
  final int? state;
  final int? pressCount;
  final double? value;
  final double? x;
  final double? y;
  final int? hatValue;

  @override
  bool operator ==(Object other) =>
      other is InputEvent &&
      controlId == other.controlId &&
      kind == other.kind &&
      flags == other.flags &&
      state == other.state &&
      pressCount == other.pressCount &&
      value == other.value &&
      x == other.x &&
      y == other.y &&
      hatValue == other.hatValue;

  @override
  int get hashCode => Object.hash(
        controlId,
        kind,
        flags,
        state,
        pressCount,
        value,
        x,
        y,
        hatValue,
      );
}

class InputEventCodec {
  static const int _flagsMask = inputEventFlagStateChanged | inputEventFlagInitial;

  static Uint8List encode(InputEvent event) {
    final writer = PayloadWriter();
    encodeTo(event, writer);
    return writer.toBytes();
  }

  static void encodeTo(InputEvent event, PayloadWriter writer) {
    final controlIdBytes = utf8.encode(event.controlId);
    if (controlIdBytes.isEmpty ||
        controlIdBytes.length > inputEventMaxControlIdLength) {
      throw const ProtocolException('controlId must be 1..64 UTF-8 bytes.');
    }
    if (!_isKnownKind(event.kind)) {
      throw const ProtocolException('Unknown input event kind.');
    }
    if ((event.flags & ~_flagsMask) != 0) {
      throw const ProtocolException('Invalid input event flags.');
    }

    writer
      ..writeUInt8(controlIdBytes.length)
      ..writeBytes(controlIdBytes)
      ..writeUInt8(event.kind)
      ..writeUInt8(event.flags);

    switch (event.kind) {
      case inputEventKindButton:
        if (event.state == null ||
            (event.state != inputEventStateUp && event.state != inputEventStateDown)) {
          throw const ProtocolException('Button event requires a valid state.');
        }
        if (event.pressCount == null) {
          throw const ProtocolException('Button event requires a pressCount.');
        }
        writer
          ..writeUInt8(event.state!)
          ..writeUInt16(event.pressCount!);
        break;
      case inputEventKindAxis:
      case inputEventKindTrigger:
        if (event.value == null) {
          throw const ProtocolException('Axis/trigger event requires a value.');
        }
        _writeRangeCheckedFloat(writer, event.value!, 0, 1, 'axis/trigger value');
        break;
      case inputEventKindStick:
        if (event.x == null || event.y == null) {
          throw const ProtocolException('Stick event requires x and y.');
        }
        _writeRangeCheckedFloat(writer, event.x!, -1, 1, 'stick x');
        _writeRangeCheckedFloat(writer, event.y!, -1, 1, 'stick y');
        break;
      case inputEventKindHat:
        if (event.hatValue == null || event.hatValue! > 8) {
          throw const ProtocolException('Hat event requires a value in 0..8.');
        }
        writer.writeUInt8(event.hatValue!);
        break;
    }
  }

  static InputEvent decode(List<int> bytes) {
    final reader = PayloadReader(bytes);
    final event = decodeFrom(reader);
    reader.expectEnd();
    return event;
  }

  static InputEvent decodeFrom(PayloadReader reader) {
    final controlIdLength = reader.readUInt8();
    if (controlIdLength < inputEventMinControlIdLength ||
        controlIdLength > inputEventMaxControlIdLength) {
      throw const ProtocolException('controlId must be 1..64 UTF-8 bytes.');
    }
    final controlId = reader.readStringOfLength(controlIdLength);
    final kind = reader.readUInt8();
    if (!_isKnownKind(kind)) {
      throw const ProtocolException('Unknown input event kind.');
    }
    final flags = reader.readUInt8();
    if ((flags & ~_flagsMask) != 0) {
      throw const ProtocolException('Invalid input event flags.');
    }

    int? state;
    int? pressCount;
    double? value;
    double? x;
    double? y;
    int? hatValue;

    switch (kind) {
      case inputEventKindButton:
        state = reader.readUInt8();
        if (state != inputEventStateUp && state != inputEventStateDown) {
          throw const ProtocolException('Invalid button state.');
        }
        pressCount = reader.readUInt16();
        break;
      case inputEventKindAxis:
      case inputEventKindTrigger:
        value = _readRangeCheckedFloat(reader, 0, 1, 'axis/trigger value');
        break;
      case inputEventKindStick:
        x = _readRangeCheckedFloat(reader, -1, 1, 'stick x');
        y = _readRangeCheckedFloat(reader, -1, 1, 'stick y');
        break;
      case inputEventKindHat:
        hatValue = reader.readUInt8();
        if (hatValue > 8) {
          throw const ProtocolException('Invalid hat value.');
        }
        break;
    }

    return InputEvent(
      controlId: controlId,
      kind: kind,
      flags: flags,
      state: state,
      pressCount: pressCount,
      value: value,
      x: x,
      y: y,
      hatValue: hatValue,
    );
  }

  static void _writeRangeCheckedFloat(
      PayloadWriter writer, double v, double min, double max, String name) {
    if (v.isNaN || v.isInfinite) {
      throw ProtocolException('$name must be finite.');
    }
    if (v < min || v > max) {
      throw ProtocolException('$name is out of range.');
    }
    writer.writeFloat32(v);
  }

  static double _readRangeCheckedFloat(
      PayloadReader reader, double min, double max, String name) {
    final v = reader.readFloat32();
    if (v.isNaN || v.isInfinite) {
      throw ProtocolException('$name must be finite.');
    }
    if (v < min || v > max) {
      throw ProtocolException('$name is out of range.');
    }
    return v;
  }

  static bool _isKnownKind(int kind) =>
      kind == inputEventKindButton ||
      kind == inputEventKindAxis ||
      kind == inputEventKindStick ||
      kind == inputEventKindTrigger ||
      kind == inputEventKindHat;
}

File: desktop/Protocol/InputEvent.cs
using System.Text;

namespace CTRL.Desktop.Protocol;

public sealed record InputEvent(
    string ControlId,
    byte Kind,
    byte Flags,
    byte? State = null,
    ushort? PressCount = null,
    float? Value = null,
    float? X = null,
    float? Y = null,
    byte? HatValue = null);

public static class InputEventCodec
{
    public const byte KindButton = 0x00;
    public const byte KindAxis = 0x01;
    public const byte KindStick = 0x02;
    public const byte KindTrigger = 0x03;
    public const byte KindHat = 0x04;

    public const byte FlagStateChanged = 0x01;
    public const byte FlagInitial = 0x02;

    public const byte StateUp = 0x00;
    public const byte StateDown = 0x01;

    public const byte MinControlIdLength = 1;
    public const byte MaxControlIdLength = 64;

    private const byte FlagsMask = FlagStateChanged | FlagInitial;

    public static byte[] Encode(InputEvent e)
    {
        var writer = new PayloadWriter();
        EncodeTo(e, writer);
        return writer.ToArray();
    }

    public static void EncodeTo(InputEvent e, PayloadWriter writer)
    {
        var controlIdBytes = StrictUtf8.GetBytes(e.ControlId);
        if (controlIdBytes.Length is < MinControlIdLength or > MaxControlIdLength)
            throw new ProtocolException("controlId must be 1..64 UTF-8 bytes.");
        if (!IsKnownKind(e.Kind))
            throw new ProtocolException("Unknown input event kind.");
        if ((e.Flags & ~FlagsMask) != 0)
            throw new ProtocolException("Invalid input event flags.");

        writer.WriteUInt8((byte)controlIdBytes.Length);
        writer.WriteBytes(controlIdBytes);
        writer.WriteUInt8(e.Kind);
        writer.WriteUInt8(e.Flags);

        switch (e.Kind)
        {
            case KindButton:
                if (e.State is null || (e.State != StateUp && e.State != StateDown))
                    throw new ProtocolException("Button event requires a valid state.");
                if (e.PressCount is null)
                    throw new ProtocolException("Button event requires a pressCount.");
                writer.WriteUInt8(e.State.Value);
                writer.WriteUInt16(e.PressCount.Value);
                break;
            case KindAxis:
            case KindTrigger:
                if (e.Value is null)
                    throw new ProtocolException("Axis/trigger event requires a value.");
                WriteRangeCheckedFloat(writer, e.Value.Value, 0f, 1f, "axis/trigger value");
                break;
            case KindStick:
                if (e.X is null || e.Y is null)
                    throw new ProtocolException("Stick event requires x and y.");
                WriteRangeCheckedFloat(writer, e.X.Value, -1f, 1f, "stick x");
                WriteRangeCheckedFloat(writer, e.Y.Value, -1f, 1f, "stick y");
                break;
            case KindHat:
                if (e.HatValue is null or > 8)
                    throw new ProtocolException("Hat event requires a value in 0..8.");
                writer.WriteUInt8(e.HatValue.Value);
                break;
        }
    }

    public static InputEvent Decode(byte[] payload)
    {
        var reader = new PayloadReader(payload);
        var e = DecodeFrom(reader);
        reader.ExpectEnd();
        return e;
    }

    public static InputEvent DecodeFrom(PayloadReader reader)
    {
        var controlIdLength = reader.ReadUInt8();
        if (controlIdLength is < MinControlIdLength or > MaxControlIdLength)
            throw new ProtocolException("controlId must be 1..64 UTF-8 bytes.");
        var controlIdBytes = reader.ReadBytes(controlIdLength);
        var kind = reader.ReadUInt8();
        if (!IsKnownKind(kind))
            throw new ProtocolException("Unknown input event kind.");
        var flags = reader.ReadUInt8();
        if ((flags & ~FlagsMask) != 0)
            throw new ProtocolException("Invalid input event flags.");

        string controlId;
        try
        {
            controlId = StrictUtf8.GetString(controlIdBytes);
        }
        catch (DecoderFallbackException)
        {
            throw new ProtocolException("Invalid UTF-8 controlId.");
        }

        byte? state = null;
        ushort? pressCount = null;
        float? value = null;
        float? x = null;
        float? y = null;
        byte? hatValue = null;

        switch (kind)
        {
            case KindButton:
                state = reader.ReadUInt8();
                if (state is not (StateUp or StateDown))
                    throw new ProtocolException("Invalid button state.");
                pressCount = reader.ReadUInt16();
                break;
            case KindAxis:
            case KindTrigger:
                value = ReadRangeCheckedFloat(reader, 0f, 1f, "axis/trigger value");
                break;
            case KindStick:
                x = ReadRangeCheckedFloat(reader, -1f, 1f, "stick x");
                y = ReadRangeCheckedFloat(reader, -1f, 1f, "stick y");
                break;
            case KindHat:
                hatValue = reader.ReadUInt8();
                if (hatValue > 8)
                    throw new ProtocolException("Invalid hat value.");
                break;
        }

        return new InputEvent(controlId, kind, flags, state, pressCount, value, x, y, hatValue);
    }

    private static void WriteRangeCheckedFloat(
        PayloadWriter writer, float v, float min, float max, string name)
    {
        if (float.IsNaN(v) || float.IsInfinity(v))
            throw new ProtocolException($"{name} must be finite.");
        if (v < min || v > max)
            throw new ProtocolException($"{name} is out of range.");
        writer.WriteFloat32(v);
    }

    private static float ReadRangeCheckedFloat(
        PayloadReader reader, float min, float max, string name)
    {
        var v = reader.ReadFloat32();
        if (float.IsNaN(v) || float.IsInfinity(v))
            throw new ProtocolException($"{name} must be finite.");
        if (v < min || v > max)
            throw new ProtocolException($"{name} is out of range.");
        return v;
    }

    private static bool IsKnownKind(byte kind) =>
        kind is KindButton or KindAxis or KindStick or KindTrigger or KindHat;

    private static readonly UTF8Encoding StrictUtf8 =
        new(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);
}
